Add nodeMetrics expapi/v2 endpoint

mvdan opened the pull request:
@vishh PTAL

vishh commented on expapi/v2/metrics_handlers.go:
```diff
@@ -0,0 +1,190 @@
+		Consumes("*/*").
+		Produces(restful.MIME_JSON)
+
+	ws.Route(ws.GET("/nodeMetrics/available/").
```
Why do we need a separate endpoint for available? Why not display the available sub resources for `/nodeMetrics/`?

vishh replied:
May be I misunderstood the purpose. Is this API returning all the available metrics like `cpu`, `memory`, etc? If yes, why do we need it?

mvdan replied:
It was just copied over from the model. As a means to check if the model was running. I'll remove it.

vishh commented on expapi/v2/metrics_handlers.go:
```diff
@@ -0,0 +1,190 @@
+	if err != nil {
+		return nil, err
+	}
+	const minute, hour, day = "1m", "1h", "1d"
```
I assume a big chunk of this function will be common to other type of metrics as well. Why make it local to node metrics?

mvdan replied:
Sure.

BTW, this is a WIP, so don't review yet as this is still being written.

vishh commented on expapi/v2/types/metrics_types.go:
```diff
@@ -0,0 +1,73 @@
+	v1.TypeMeta   `json:",inline"`
+	v1.ObjectMeta `json:"metadata,omitempty"`
+
+	NodeMetrics      MetricsWindows
```
json tag for this?

vishh commented on expapi/v2/types/metrics_types.go:
```diff
@@ -0,0 +1,73 @@
+
+	NodeMetrics      MetricsWindows
+	SystemContainers []struct {
+		ContainerName string         `json:"containername"`
```
nit: Let's stick to camelCase starting with lower case for json tags.

vishh commented on expapi/v2/types/metrics_types.go:
```diff
@@ -0,0 +1,73 @@
+}
+
+type DerivedNodeMetricsList struct {
+	v1.TypeMeta `json:",inline"`
```
Are the `meta` objects required from heapster perspective? Since it provides read-only APIs I don't see a reason. I could be wrong.

mvdan replied:
We can remove them.

vishh commented on expapi/v2/metrics_handlers.go:
```diff
@@ -0,0 +1,190 @@
+func (a *Api) RegisterMetrics(container *restful.Container) {
+	ws := new(restful.WebService)
+	ws.
+		Path("/experimental/v2").
```
We should specify that `experimental` means it will be deleted anytime. Also why not use `experimental/v0.1` ?

mvdan replied:
It's `experimental/v2` since this may become `api/v2`.

vishh replied:
ok

On Tue, Sep 15, 2015 at 1:25 PM, Daniel Martí <notifications@github.com>
wrote:

> In expapi/v2/metrics_handlers.go
> <https://github.com/kubernetes/heapster/pull/549#discussion_r39560080>:
>
> > +	model_api "k8s.io/heapster/model"
> > +	"k8s.io/heapster/store/statstore"
> > +	"k8s.io/kubernetes/pkg/api/resource"
> > +)
> > +
> > +// errModelNotActivated is the error that is returned when manager.cluster
> > +// has not beed initialized.
> > +var errModelNotActivated = errors.New("the model is not activated")
> > +
> > +// RegisterMetrics registers the Metrics API endpoints.
> > +// All endpoints that end with a {metric-name} also receive a start time query parameter.
> > +// The start and end times should be specified as a string, formatted according to RFC 3339.
> > +func (a *Api) RegisterMetrics(container *restful.Container) {
> > +	ws := new(restful.WebService)
> > +	ws.
> > +		Path("/experimental/v2").
>
> It's experimental/v2 since this may become api/v2.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/heapster/pull/549/files#r39560080>.
>

vishh commented:
Just some nits. I will make another pass once the PR is ready for review. Overall LGTM!

mvdan commented:
@vishh comments amended and WIP removed. Checked that the endpoint works. Note the two TODOs.

vishh commented on model/types.go:
```diff
@@ -48,12 +48,12 @@ type Model interface {
-	GetPodStats(PodRequest) (map[string]StatBundle, time.Duration, error)
-	GetPodContainerStats(PodContainerRequest) (map[string]StatBundle, time.Duration, error)
-	GetFreeContainerStats(FreeContainerRequest) (map[string]StatBundle, time.Duration, error)
+	GetClusterStats() (map[string]StatBundle, time.Time, time.Duration, error)
```
nit: We should have explicit names for the return arguments or at-least document their meaning.

vishh replied:
Also the returns types are essentially the same. Is it worth unifying them into a single struct and re-use in all methods?

mvdan replied:
Yes.

mvdan commented:
@vishh just applied your stats result struct suggestion, makes everything a lot simpler :)

There's only one TODO left, the system containers bit.

mvdan commented:
Re-adding WIP until the TODO is fixed.

mvdan commented:
@vishh progress update:

* Rebased on master
* Updated the experimental api types to match the ones in nodeMetrics (replace map with list et al)
* Updated kubernetes in godeps to my nodeMetrics PR
* Add kube-heapster-bridge

vishh commented:
@mvdan: is this still a WIP?

mvdan commented:
FWIW, I'll keep WIP since I'm pushing my progress to this PR.

mvdan commented:
@vishh rebased on master and removed the bridge changes - they'll come in a separate PR. Ready to merge.

vishh commented on expapi/v2/metrics_handlers.go:
```diff
@@ -0,0 +1,185 @@
+func (a *Api) RegisterMetrics(container *restful.Container) {
+	ws := new(restful.WebService)
+	ws.
+		Path("/experimental/v2").
```
What is the output of this endpoint?

mvdan replied:
This isn't an endpoint - it's the path for the whole WebService. Basically, the prefix and other common parameters for the endpoints under it. So `/nodeMetrics` below becomes `/experimental/v2/nodeMetrics`, etcetera.

vishh commented:
LGTM
